File: ocaml/plugins/gcpsecrets.ml
(* The `gcpsecrets` provider kind, as a voxgig/plugin definition.

   Google Secret Manager over HTTPS, with the metadata server as the
   fallback source of a token.

   A port of typescript/plugins/gcpsecrets.ts, which is canonical. *)

open Secret
open Provider
open Httpjson

(* GCP Secret Manager.

   `api.token` reads secret `api_token` (dots flattened to `_`; Secret
   Manager ids have no hierarchy and reject dots), latest version. The token
   comes from config, then GOOGLE_OAUTH_ACCESS_TOKEN, then the GCE/GKE
   metadata server - so on Google's own platform no credential configuration
   is needed at all.

   The metadata call itself is plain http to a link-local host by platform
   design and carries no credential, so `checkaddr` guards the Secret
   Manager address instead. *)
let gcpsecrets_provider (project : string) (token : string) (addr : string)
    (metadataaddr : string) : provider =
  let livetoken = ref None in
  let renewat = ref never in

  let usemetadataaddr () =
    if "" <> metadataaddr then metadataaddr
    else
      let host = getenv "GCE_METADATA_HOST" in
      if "" <> host then "http://" ^ host else "http://metadata.google.internal"
  in

  let login () =
    let configured = firstof [ token; getenv "GOOGLE_OAUTH_ACCESS_TOKEN" ] in
    if "" <> configured then configured
    else begin
      let url =
        trimslash (usemetadataaddr ()) ^ "/computeMetadata/v1/instance/service-accounts/default/token"
      in

      let res = fetchjson ~headers:[ ("Metadata-Flavor", "Google") ] "GET" url in
      let got = digtext res.jbody [ "access_token" ] in

      if 200 <> res.status || not (isset got) then
        fail "sekreto: gcp: no token and metadata server did not answer";

      renewat := renewtime (Json.odig res.jbody [ "expires_in" ]);
      Option.get got
    end
  in

  {
    lookup =
      (fun name ->
        if "" = project then fail "sekreto: gcp: no project";

        let useaddr = firstof [ addr; "https://secretmanager.googleapis.com" ] in
        checkaddr useaddr;

        if None = !livetoken || nowms () >= !renewat then livetoken := Some (login ());

        let url =
          trimslash useaddr ^ "/v1/projects/" ^ project ^ "/secrets/" ^ flatname name "_"
          ^ "/versions/latest:access"
        in

        let res =
          fetchjson
            ~headers:[ ("authorization", "Bearer " ^ Option.value ~default:"" !livetoken) ]
            "GET" url
        in

        if 404 = res.status then None
        else if 200 <> res.status then
          fail ("sekreto: gcp error: " ^ string_of_int res.status ^ ": " ^ url)
        else
          match digstr res.jbody [ "payload"; "data" ] with
          | None -> None
          | Some data -> (
            match Http.unbase64 data with
            | Some bytes -> Some bytes
            | None -> fail "sekreto: gcp: undecodable secret"));
    describe = (fun () -> "gcpsecrets:" ^ project);
  }

let plugin () : Defs.definition =
  providerplugin "gcpsecrets" (fun spec ->
      gcpsecrets_provider spec.project spec.token spec.addr spec.metadataaddr)

File: elixir/plugins/gcpsecrets.ex
# GCP Secret Manager, as a sekreto plugin.
#
# A port of typescript/plugins/gcpsecrets.ts, which is canonical.

defmodule Sekreto.Plugins.Gcpsecrets do
  @moduledoc """
  The `gcpsecrets` provider kind.

  A plugin module: nothing under `src/` names it, and a chain that does not
  name this kind never loads it. See docs/design/plugin-providers.md.
  """

  alias Sekreto.Cell
  alias Sekreto.Error
  alias Sekreto.Json
  alias Sekreto.Plugins.Httpjson
  alias Sekreto.ProviderSpec
  alias Sekreto.Providers

  @doc """
  The `gcpsecrets` provider kind, as a voxgig/plugin definition.

  The calling project passes it to the constructor, and only then can a
  chain name the kind:

      Sekreto.new(chain, plugins: [Sekreto.Plugins.Gcpsecrets.gcpsecrets()])
  """
  def gcpsecrets, do: Providers.providerplugin("gcpsecrets", &fromspec/1)

  @doc """
  GCP Secret Manager.

  `api.token` reads secret `api_token` (dots flattened to `_`; Secret
  Manager ids have no hierarchy and reject dots), latest version. The token
  comes from config, then `GOOGLE_OAUTH_ACCESS_TOKEN`, then the GCE/GKE
  metadata server - so on Google's own platform no credential configuration
  is needed at all.

  The metadata call itself is plain http to a link-local host by platform
  design; no credential rides on it, so `checkaddr` guards the Secret
  Manager address instead.
  """
  def gcpsecrets(project, token, addr, metadataaddr) do
    cell = Cell.new(%{token: "", renewat: Httpjson.never()})

    usemetadataaddr = fn ->
      cond do
        "" != metadataaddr -> metadataaddr
        "" != Providers.getenv("GCE_METADATA_HOST") ->
          "http://" <> Providers.getenv("GCE_METADATA_HOST")
        true -> "http://metadata.google.internal"
      end
    end

    login = fn ->
      configured = Providers.first([token, Providers.getenv("GOOGLE_OAUTH_ACCESS_TOKEN")])

      if "" != configured do
        Cell.put(cell, %{token: configured, renewat: Httpjson.never()})
      else
        url =
          Httpjson.trimslash(usemetadataaddr.()) <>
            "/computeMetadata/v1/instance/service-accounts/default/token"

        res = Httpjson.fetchjson("GET", url, [{"Metadata-Flavor", "Google"}])
        got = Json.text(Json.dig(res.body, ["access_token"]))

        if 200 != res.status or :none == got or "" == got do
          raise Error, message: "sekreto: gcp: no token and metadata server did not answer"
        end

        renewat = Httpjson.renewtime(Json.dig(res.body, ["expires_in"]))
        Cell.put(cell, %{token: got, renewat: renewat})
      end
    end

    %{
      lookup: fn name ->
        if "" == project, do: raise(Error, message: "sekreto: gcp: no project")

        useaddr = Providers.first([addr, "https://secretmanager.googleapis.com"])
        Providers.checkaddr(useaddr)

        if Httpjson.expired?(cell), do: login.()

        url =
          Httpjson.trimslash(useaddr) <>
            "/v1/projects/" <>
            project <> "/secrets/" <> Sekreto.flatname(name, "_") <> "/versions/latest:access"

        res =
          Httpjson.fetchjson("GET", url, [{"authorization", "Bearer " <> Cell.get(cell).token}])

        cond do
          404 == res.status ->
            nil

          200 != res.status ->
            raise Error,
              message: "sekreto: gcp error: " <> Httpjson.tostr(res.status) <> ": " <> url

          true ->
            case Json.asstr(Json.dig(res.body, ["payload", "data"])) do
              :none ->
                nil

              data ->
                case Httpjson.unbase64(data) do
                  :error -> raise Error, message: "sekreto: gcp: undecodable secret"
                  text -> text
                end
            end
        end
      end,
      describe: fn -> "gcpsecrets:" <> project end
    }
  end

  defp fromspec(%ProviderSpec{} = spec),
    do: gcpsecrets(spec.project, spec.token, spec.addr, spec.metadataaddr)
end
